pkg/model/model.go

```diff
@@ -0,0 +1,295 @@
+	Channels       map[string]*Channel
+}
+
+func (m *Package) Validate() error {
+	var result *multierror.Error
+	if m.Name == "" {
+		result = multierror.Append(result, errors.New("package name must not be empty"))
+	}
+
+	if err := m.Icon.Validate(); err != nil {
+		result = multierror.Append(result, fmt.Errorf("invalid icon: %v", err))
+	}
+
+	if len(m.Channels) == 0 {
+		result = multierror.Append(result, fmt.Errorf("package must contain at least one channel"))
+	}
+
+	if m.DefaultChannel == nil {
+		result = multierror.Append(result, fmt.Errorf("default channel must be set"))
+	}
+
+	foundDefault := false
+	for name, ch := range m.Channels {
+		if name != ch.Name {
+			result = multierror.Append(result, fmt.Errorf("channel key %q does not match channel name %q", name, ch.Name))
+		}
+		if err := ch.Validate(); err != nil {
+			result = multierror.Append(result, fmt.Errorf("invalid channel %q: %v", ch.Name, err))
+		}
+		if ch == m.DefaultChannel {
+			foundDefault = true
+		}
+		if ch.Package != m {
+			result = multierror.Append(result, fmt.Errorf("channel %q not correctly linked to parent package", ch.Name))
+		}
+	}
+
+	if m.DefaultChannel != nil && !foundDefault {
+		result = multierror.Append(result, fmt.Errorf("default channel %q not found in channels list", m.DefaultChannel.Name))
+	}
+	return result.ErrorOrNil()
+}
+
+type Icon struct {
+	Data      []byte
+	MediaType string
+}
+
+func (i *Icon) Validate() error {
+	if i == nil {
+		return nil
+	}
+	// TODO(joelanford): Should we check that data and mediatype are set,
+	//   and detect the media type of the data and compare it to the
+	//   mediatype listed in the icon field? Currently, some production
+	//   index databases are failing these tests, so leaving this
+	//   commented out for now.
+	var result *multierror.Error
+	//if len(i.Data) == 0 {
+	//	result = multierror.Append(result, errors.New("icon data must be set if icon is defined"))
+	//}
+	//if len(i.MediaType) == 0 {
+	//	result = multierror.Append(result, errors.New("icon mediatype must be set if icon is defined"))
+	//}
+	//if len(i.Data) > 0 {
+	//	if err := i.validateData(); err != nil {
+	//		result = multierror.Append(result, err)
+	//	}
+	//}
+	return result.ErrorOrNil()
+}
+
+func (i *Icon) validateData() error {
+	if !filetype.IsImage(i.Data) {
+		return errors.New("icon data is not an image")
+	}
+	t, err := filetype.Match(i.Data)
+	if err != nil {
+		return err
+	}
+	if t.MIME.Value != i.MediaType {
+		return fmt.Errorf("icon media type %q does not match detected media type %q", i.MediaType, t.MIME.Value)
+	}
+	return nil
+}
+
+type Channel struct {
+	Package *Package
+	Name    string
+	Bundles map[string]*Bundle
+}
+
+// TODO(joelanford): This function determines the channel head by finding the bundle that has 0
+//   incoming edges, based on replaces and skips. It also expects to find exactly one such bundle.
+//   Is this the correct algorithm?
+func (c Channel) Head() (*Bundle, error) {
```

Comment:
It seems like adding `substitutes-for` support in declarative config is unnecessary complexity. In declarative config, the upgrade graph can be edited at any point in a process that produces or updates the final index image.

In declarative-config world, inserting a new bundle that substitutes for another bundle is fairly easy:
1. Add a new `olm.bundle` blobs for (using that EP as the example) `1.0.1-patched`, with channel properties that include `replaces: 1.0.0` for each relevant channel.
2. Find all `olm.bundle` blobs that contain a channel property with `replaces: 1.0.0` in the relevant channels, and update those properties to `replaces: 1.0.1-patched`

No changes to the bundle images necessary.

Reply:
Wouldn't the substitution problem go away with declarative configs because the update graph is explicitly defined? I think this is addressed in the [alternatives section](https://github.com/operator-framework/enhancements/blob/master/enhancements/substitutes-for.md#declarative-config).

Reply:
I agree, Kevin. So it sounds like we're saying:
1. substitutes-for is a way to tell tooling how to build upgrade edges in an index.
2. the upgrade edge primitives that exist today in declarative config are `olm.skips` and `olm.channel` (with replaces) properties.
3. Index maintainers that wish to consider `substitutes-for` for edge generation in declarative configs will use a tool that considers `substitutes-for` values when manipulating `olm.skips` and `olm.channel` properties in `olm.bundle` blobs.

Does that sound right?